### PR TITLE
feat(model3d): Basic React controls

### PR DESCRIPTION
**Pull request by ConradJChan**
Initial React rewrite of Model3DViewer controls

TODO:

- [x] Unit tests

Before:
Screen Shot 2021-05-03 at 11 12 40 AM

After:
Screen Shot 2021-05-03 at 11 12 56 AM

### CODE REVIEW

File: src/lib/Preview.js [hunk 1]
```diff
@@ -995,6 +995,8 @@ class Preview extends EventEmitter {
         // Add the response interceptor to the preview instance
         this.options.responseInterceptor = options.responseInterceptor;
 
+        this.options.useReactControlsModel3D = !!options.useReactControlsModel3D;
```
Comment (jstoffan) [inline src/lib/Preview.js:998]: Can we use a viewer option similar to the media controls?

File: src/lib/viewers/box3d/Box3DViewer.js [hunk 2]
```diff
@@ -167,7 +168,7 @@ class Box3DViewer extends BaseViewer {
      * @return {void}
      */
     destroySubModules() {
-        if (this.controls) {
+        if (this.controls && !this.options.useReactControlsModel3D) {
             this.controls.destroy();
```
Comment (jstoffan) [inline src/lib/viewers/box3d/Box3DViewer.js:172]: Do we still need to call `destroy` on the controls root?

File: src/lib/viewers/box3d/model3d/Model3DViewer.js [hunk 3]
```diff
@@ -28,19 +31,27 @@ const LOAD_TIMEOUT = 180000; // 3 minutes
...
         forward: null,
     };
 
+    useReactControls = false;
+
     /** @inheritdoc */
     constructor(option) {
         super(option);
 
+        this.useReactControls = !!this.options.useReactControlsModel3D;
```
Comment (jstoffan) [inline src/lib/viewers/box3d/model3d/Model3DViewer.js:53]: Do we need a second copy of this value if it's already available on `this.options`?

File: src/lib/viewers/box3d/model3d/Model3DViewer.js [hunk 4]
```diff
@@ -239,15 +256,31 @@ class Model3DViewer extends Box3DViewer {
...
+                    const { name, start, stop } = animations[0].getClip(clipId);
+                    const duration = stop - start;
+                    return {
+                        duration,
+                        id: clipId,
+                        name,
+                    };
+                });
+
+                this.renderer.setAnimationClip(this.animationClips[0].id);
```
Comment (jstoffan) [inline src/lib/viewers/box3d/model3d/Model3DViewer.js:270]: Should we be acting on `renderer` inside of a controls-specific method?

File: src/lib/viewers/controls/model3d/AnimationClipsToggle.scss [hunk 5]
```diff
@@ -0,0 +1,5 @@
+@import 'src/lib/viewers/controls/styles';
```
Comment (jstoffan) [inline src/lib/viewers/controls/model3d/AnimationClipsToggle.scss:1]: Can we use a relative path here?

File: src/lib/viewers/controls/model3d/ResetControl.tsx [hunk 6]
```diff
@@ -0,0 +1,21 @@
+import React from 'react';
+import IconReset24 from '../icons/IconReset24';
+import './ResetControl.scss';
+
+export type Props = {
+    onReset: () => void;
+};
+
+export default function ResetControl({ onReset }: Props): JSX.Element {
+    const title = __('box3d_reset');
```
Comment (jstoffan) [inline src/lib/viewers/controls/model3d/ResetControl.tsx:10]: Can this be inlined?

File: src/lib/viewers/box3d/model3d/Model3DViewer.js (hunk 4 above)
Comment (ConradJChan) [reply src/lib/viewers/box3d/model3d/Model3DViewer.js:270]: Ideally not, but it was being done previously in a more roundabout way -- L282 sets the animation clip => Model3DControls would emit a `EVENT_SELECT_ANIMATION_CLIP` event => `handleSelectAnimationClip` which acts on the `renderer`

File: src/lib/viewers/box3d/model3d/Model3DViewer.js [hunk 7]
```diff
@@ -301,6 +347,10 @@ class Model3DViewer extends Box3DViewer {
             this.renderer.stopAnimation();
             this.renderer.resetView();
         }
+
+        if (this.controls && this.useReactControls) {
```
[the remark this answers is not in the document]
Comment (ConradJChan) [reply src/lib/viewers/box3d/model3d/Model3DViewer.js:351]: No, it should be separate, are you thinking I should move this up to be located closer to the controls block?

File: src/lib/viewers/controls/model3d/ResetControl.scss [hunk 8]
```diff
@@ -0,0 +1,5 @@
+@import 'src/lib/viewers/controls/styles';
```
Comment (jstoffan) [inline src/lib/viewers/controls/model3d/ResetControl.scss:1]: Can we use a relative path here, as well?

File: src/lib/viewers/controls/icons/IconAnimation24.tsx [hunk 9]
```diff
@@ -0,0 +1,18 @@
+import * as React from 'react';
+
+function IconAnimation24(props: React.SVGProps<SVGSVGElement>): JSX.Element {
+    return (
+        <svg focusable={false} height={24} width={24} {...props}>
+            <g fill="none" fillRule="evenodd">
```
Comment (jstoffan) [inline src/lib/viewers/controls/icons/IconAnimation24.tsx:6]: Can we remove this wrapper?

File: src/lib/viewers/controls/icons/IconAnimation24.tsx [hunk 10]
```diff
@@ -0,0 +1,18 @@
...
+                <path d="M0-1h24v24H0" />
```
Comment (jstoffan) [inline src/lib/viewers/controls/icons/IconAnimation24.tsx:7]: Can this be removed?

**Review by jstoffan — APPROVED** [review]

File: src/lib/viewers/box3d/model3d/Model3DViewer.js [hunk 11]
```diff
@@ -28,9 +31,13 @@ const LOAD_TIMEOUT = 180000; // 3 minutes
  * This is the entry point for the model3d preview.
  */
 class Model3DViewer extends Box3DViewer {
+    animationClips = [];
+
     /** @property {Object[]} - List of Box3D instances added to the scene */
     instances = [];
 
+    isAnimationPlaying = false;
```
Comment (jstoffan) [inline src/lib/viewers/box3d/model3d/Model3DViewer.js:39]: Should we add JSDocs to these for consistency within the file?

File: src/lib/viewers/box3d/model3d/__tests__/Model3DControlsNew-test.tsx [hunk 12]
```diff
@@ -0,0 +1,41 @@
...
+                    currentAnimationClipId="123"
+                    isPlaying={false}
+                    onAnimationClipSelect={onAnimationClipSelect}
+                    onFullscreenToggle={onFullscreenToggle}
+                    onPlayPause={onPlayPause}
+                    onReset={onReset}
+                />,
+            );
+
+            wrapper.find(ResetControl).prop('onReset')();
```
Comment (jstoffan) [inline src/lib/viewers/box3d/model3d/__tests__/Model3DControlsNew-test.tsx:28]: Should this be in a separate test?

File: src/lib/viewers/box3d/model3d/__tests__/Model3DViewer-test.js [hunk 13]
```diff
@@ -247,6 +285,17 @@ describe('lib/viewers/box3d/model3d/Model3DViewer', () => {
                     });
                 });
             });
+
+            describe('with reactcontrols enabled', () => {
```
Comment (jstoffan) [inline src/lib/viewers/box3d/model3d/__tests__/Model3DViewer-test.js:289]: Looks like this is missing a space.

**Review by jstoffan — APPROVED** [review]

**Review by jstoffan — APPROVED** [review]

**Pull request merged**